browser/tor/onion_location_tab_helper.h

```diff
@@ -0,0 +1,41 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_TOR_ONION_LOCATION_TAB_HELPER_H_
+#define BRAVE_BROWSER_TOR_ONION_LOCATION_TAB_HELPER_H_
+
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+namespace tor {
+
+class OnionLocationTabHelper
+    : public content::WebContentsObserver,
+      public content::WebContentsUserData<OnionLocationTabHelper> {
+ public:
+  ~OnionLocationTabHelper() override;
+
+  static void SetOnionLocation(content::WebContents* web_contents,
+                               const GURL& onion_location);
+
+  bool should_show_icon() const { return !onion_location_.is_empty(); }
+
+  GURL onion_location() const { return onion_location_; }
+
+  WEB_CONTENTS_USER_DATA_KEY_DECL();
+
+ private:
+  explicit OnionLocationTabHelper(content::WebContents* web_contents);
+  friend class content::WebContentsUserData<OnionLocationTabHelper>;
+
+  GURL onion_location_;
```

Comment:
I guess we should clean this state if a navigation happened

Reply:
It will be cleaned up if the new navigation destination doesn't have onion-location header per web_contents
https://github.com/brave/brave-core/pull/6762/files#diff-d8f2fb41afdc92a8124ec8a331c1d44dR83